File: typescript/chapters/ch06/src/core/profiles.ts
// 章节能力快照：每章一个冻结单例，组合根用引用相等校验 profile 来源。
// P06 在 P05 基础上增加 subagent 能力，通知组合根为 Agent 注册子代理委派工具。
export type Capability =
  | "loop"
  | "powershell"
  | "tool_registry"
  | "files"
  | "policy"
  | "hooks"
  // todo 表示 Agent 已具备显式计划快照与陈旧计划提醒能力。
  | "todo"
  // subagent 表示可将自包含工作交给隔离 AgentRunner 执行。
  | "subagent";

export interface ChapterProfile {
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

// 用不可导出底层 Set 的包装实现 ReadonlySet，防止 profile 能力被外部修改。
class CapabilitySet implements ReadonlySet<Capability> {
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    this.#values = new Set(values);
  }

  get size(): number {
    return this.#values.size;
  }

  has(value: Capability): boolean {
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

// P04 在 P03 基础上加入 hooks，允许受限 Hook 生命周期扩展。
export const P04: ChapterProfile = Object.freeze({
  chapter: 4,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
  ]),
});

// P05 在 P04 基础上加入 todo，用于会话级计划快照与陈旧提醒。
export const P05: ChapterProfile = Object.freeze({
  chapter: 5,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
  ]),
});

export const P06: ChapterProfile = Object.freeze({
  // P06 在 P05 的计划能力上加入受策略约束的单次子任务委派。
  chapter: 6,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
  ]),
});

// 仅返回模块内冻结的单例，供组合根用引用相等性校验 profile 来源。
export function profileForChapter(chapter: number): ChapterProfile {
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (chapter === 4) {
    return P04;
  }
  if (chapter === 5) {
    return P05;
  }
  if (chapter === 6) {
    return P06;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}

File: python/ch05_agent/agent_ch05/core/profiles.py
"""固定章节能力快照。

Java 对照：`ChapterProfile` 类似不可变配置 record，P01-P05 类似预定义单例常量。
组合根使用对象身份判断，调用方不能临时拼一个同字段对象来冒充正式章节。

这是什么：定义每章 Agent 允许使用的能力边界
为什么需要：防止越级使用未学功能，确保教程循序渐进
"""

from dataclasses import dataclass
from typing import Literal

# 能力字面量联合类型，类似 Java 的能力枚举
Capability = Literal["loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo"]


@dataclass(frozen=True, slots=True)  # frozen=True 表示不可变，类似 Java record
class ChapterProfile:
    """章节号及其允许装配的能力白名单。

    这是什么：一个章节的能力配置对象
    Java 类比：record ChapterProfile(int chapter, Set<Capability> capabilities)
    为什么需要：让组合根通过对象身份（is 比较）验证章节合法性，防止伪造
    """
    chapter: int  # 章节序号，1-20
    capabilities: frozenset[Capability]  # 该章允许的能力集合，frozenset 是不可变 Set


# 预定义的章节单例常量，类似 Java 的 public static final ChapterProfile P01 = ...
P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))  # 第 1 章：基础循环 + Shell
P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))  # 第 2 章：+工具注册+文件
P03 = ChapterProfile(3, frozenset({"loop", "powershell", "tool_registry", "files", "policy"}))  # 第 3 章：+权限策略
P04 = ChapterProfile(4, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks"}))  # 第 4 章：+Hook 生命周期
P05 = ChapterProfile(5, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo"}))  # 第 5 章：+TODO 跟踪


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """只返回模块内固定单例；尚未迁移的章节明确报错。

    这是什么：根据章节号返回对应的能力配置
    Java 类比：static ChapterProfile forChapter(int chapter) throws IllegalArgumentException
    为什么需要：CLI 参数只能传数字，需要转换成预定义的配置对象
    """
    # 固定映射表：只有这些章节已经实现，类似 Java 的 Map.of(1, P01, 2, P02, ...)
    profiles = {1: P01, 2: P02, 3: P03, 4: P04, 5: P05}

    # Python 的 bool 是 int 的子类，必须先排除 True/False，避免 chapter=True 被当成 1
    if isinstance(chapter, bool) or not isinstance(chapter, int) or not 1 <= chapter <= 20:
        raise ValueError("chapter 必须是 1 到 20 的整数")

    # 尚未实现的章节明确拒绝，而不是返回一个临时拼凑的对象
    if chapter not in profiles:
        raise ValueError(f"第 {chapter} 章尚未迁移为 Python")

    return profiles[chapter]  # 返回模块级单例，组合根用 is 判断对象身份
